gslib/tests/test_cp.py

```diff
@@ -1002,6 +1002,42 @@ def _Check2():
 
     _Check2()
 
+  @SequentialAndParallelTransfer
+  def test_request_reason_header(self):
+    """Tests that x-goog-request-reason is set successfully on copy."""
+    bucket_uri = self.CreateBucket()
+    dst_uri = suri(bucket_uri, 'foo')
+    fpath = self._get_test_file('test.gif')
+
+    self.RunGsUtil(['-h', 'X-Goog-Request-Reason:b/this_is_reason', 'cp',
+                    fpath, dst_uri])
+    stdout = self.RunGsUtil(['ls', '-L', dst_uri], return_stdout=True)
+    self.assertRegex(stdout.lower(), r'x-goog-request-reason:b/this_is_reason')
+    dst_uri2 = suri(bucket_uri, 'bar')
+    self.RunGsUtil(['cp', dst_uri, dst_uri2])
+    # Ensure metadata was preserved across copy.
+    stdout = self.RunGsUtil(['ls', '-L', dst_uri2], return_stdout=True)
+    self.assertRegex(stdout.lower(), r'x-goog-request-reason:b/this_is_reason')
+
+  @SequentialAndParallelTransfer
+  def test_request_reason_set_by_environment_var(self):
+    """Tests that request reason can be set successfully using the environment
+    variable CLOUDSDK_CORE_REQUEST_REASON."""
+    os.environ['CLOUDSDK_CORE_REQUEST_REASON'] = 'b/this_is_env_reason'
+    bucket_uri = self.CreateBucket()
+    dst_uri = suri(bucket_uri, 'foo')
+    fpath = self._get_test_file('test.gif')
+    self.RunGsUtil(['cp', fpath, dst_uri])
```

Comment:
It turns out that the `cp` operation doesn't work under the XML API, which is why this test fails when run using XML.  (Note that the `ls` command still works, but the output is slightly different, as I mentioned below)

When I run this command with debug output enabled, it looks like gsutil is sending the header:

```
$ $ ./gsutil -D -o 'GSUtil:prefer_api=xml' -h 'x-goog-request-reason: b/1' cp <(echo testingcontents) gs://do-not-delete-lklk/testreqreasonxml
[...]
send: u'PUT /do-not-delete-lklk/testreqreasonxml HTTP/1.1\r\nHost: storage.googleapis.com\r\nAccept-Encoding: identity\r\nX-Goog-Request-Reason:  b/1\r\nExpect: 100-Continue\r\ncontent-language: en\r\nTransfer-Encoding: chunked\r\nx-goog-api-version: 1\r\nContent-Type: application/octet-stream\r\nAuthorization: <REDACTED>
[...]
```

Do we want to hold off on merging this until the XML API accepts and correctly stores this header?  If so, please point me to an internal tracking bug number for this.

Reply:
> It turns out that the `cp` operation doesn't work under the XML API, which is why this test fails when run using XML. (Note that the `ls` command still works, but the output is slightly different, as I mentioned below)
> 
> When I run this command with debug output enabled, it looks like gsutil is sending the header:
> 
> ```
> $ $ ./gsutil -D -o 'GSUtil:prefer_api=xml' -h 'x-goog-request-reason: b/1' cp <(echo testingcontents) gs://do-not-delete-lklk/testreqreasonxml
> [...]
> send: u'PUT /do-not-delete-lklk/testreqreasonxml HTTP/1.1\r\nHost: storage.googleapis.com\r\nAccept-Encoding: identity\r\nX-Goog-Request-Reason:  b/1\r\nExpect: 100-Continue\r\ncontent-language: en\r\nTransfer-Encoding: chunked\r\nx-goog-api-version: 1\r\nContent-Type: application/octet-stream\r\nAuthorization: <REDACTED>
> [...]
> ```
> 
> Do we want to hold off on merging this until the XML API accepts and correctly stores this header? If so, please point me to an internal tracking bug number for this.

We don't have another bug for just XML API, I guess I should fix it in this pull request. Actually, can you point me to the code where it is related to XML API, I have hard time to identify where to change for XML API. Thanks!

Reply:
Per our offline thread, we determined that this should **not** be setting the request reason as custom metadata on the object.  This should be used only for auth'z purposes, and not preserved in the metadata.  We should instead be testing that:

1) If this header is supplied for a `cp` command, the resulting object's custom metadata should **not** contain the request reason
2) The header should be included along with the request.  We can probably check for this by running the command in debug mode, e.g. `gsutil -D cp src dst` and checking for the header in the output of stderr.

Reply:
I think we probably want to change this PR to:

- Only accept the request reason from an environment variable
- [JSON API] Change gslib/gcs_json_api.py such that all when all requests are made, they're wrapped in a  `_ApitoolsRequestHeaders()` context manager, and that context manager method is the place where we should be checking the environment variable for a request reason and adding it to the headers if it's present.
- [XML API] Change gslib/boto_translation.py to do the same thing; the method that creates headers for this API client is `_CreateBaseHeaders()`.

Reply:
I think we probably want to change this PR to:

- Only accept the request reason from an environment variable
- [JSON API] Change gslib/gcs_json_api.py such that when we construct an api_client, the `additional_http_headers` object checks for the request reason and adds it to that dict if present
- [XML API] Change gslib/boto_translation.py to do something similar; before we send each request, we construct the base headers -- the method that creates headers for this API client is `_CreateBaseHeaders()`.  In that method, we should check for the request reason env var, and set it in the headers dict if present.

Reply:
I started a new branch with these things done; I'll leave the test writing to you :)

I'll go ahead and close this PR; to build on top of the changes I checked in, check out the `req-reason-header` branch and make your changes on top of that; push them to a branch in your fork, and send a PR to merge your changes into the `req-reason-header` branch.  Then we can merge that whole branch into master when everything is finished.

Reply:
Thank you very much, will do!